crates/transaction-pool/src/pool.rs

```diff
@@ -0,0 +1,756 @@
+
+    /// Updates the pool statistics if transaction was removed.
+    fn finalize_remove(&mut self, hash: &T::Hash) -> Option<Arc<T>> {
+        self.by_hash.remove(hash).map(|old| {
+            self.mem_usage -= old.transaction.mem_usage();
+            old.transaction
+        })
+    }
+
+    /// Updates best and worst transactions from a sender.
+    fn update_senders_worst_and_best(
+        &mut self,
+        previous: Option<((S::Score, Transaction<T>), (S::Score, Transaction<T>))>,
+        current: Option<((S::Score, Transaction<T>), (S::Score, Transaction<T>))>,
+    ) {
+        let worst_collection = &mut self.worst_transactions;
+        let best_collection = &mut self.best_transactions;
+
+        let is_same = |a: &(S::Score, Transaction<T>), b: &(S::Score, Transaction<T>)| {
+            a.0 == b.0 && a.1.hash() == b.1.hash()
+        };
+
+        let update = |collection: &mut BTreeSet<_>, (score, tx), remove| {
+            if remove {
+                collection.remove(&ScoreWithRef::new(score, tx));
+            } else {
+                collection.insert(ScoreWithRef::new(score, tx));
+            }
+        };
+
+        match (previous, current) {
+            (None, Some((worst, best))) => {
+                update(worst_collection, worst, false);
+                update(best_collection, best, false);
+            }
+            (Some((worst, best)), None) => {
+                // all transactions from that sender has been removed.
+                // We can clear a hashmap entry.
+                self.transactions.remove(worst.1.sender());
+                update(worst_collection, worst, true);
+                update(best_collection, best, true);
+            }
+            (Some((w1, b1)), Some((w2, b2))) => {
+                if !is_same(&w1, &w2) {
+                    update(worst_collection, w1, true);
+                    update(worst_collection, w2, false);
+                }
+                if !is_same(&b1, &b2) {
+                    update(best_collection, b1, true);
+                    update(best_collection, b2, false);
+                }
+            }
+            (None, None) => {}
+        }
+    }
+
+    /// Attempts to remove the worst transaction from the pool if it's worse than the given one.
+    ///
+    /// Returns `None` in case we couldn't decide if the transaction should replace the worst transaction or not.
+    /// In such case we will accept the transaction even though it is going to exceed the limit.
+    fn remove_worst(
+        &mut self,
+        transaction: &Transaction<T>,
+        replace: &dyn ShouldReplace<T>,
+    ) -> error::Result<Option<Transaction<T>>, T::Hash> {
+        let to_remove = match self.worst_transactions.iter().next_back() {
+            // No elements to remove? and the pool is still full?
+            None => {
+                warn!("The pool is full but there are no transactions to remove.");
+                return Err(error::Error::TooCheapToEnter(
```

Comment:
Is `TooCheapToEnter` the right error type here? Same in line 340.

Reply:
The original author probably had a wider context on his mind when deciding on the errors. In wider context, function "remove_worst" is called only on the import of the new function. So, if the imported transaction is too cheap, then no transaction should be removed from the pool.

Reply:
Makes sense.